File: 2024-4-17/dungeon game.py
class Solution:
    def calculateMinimumHP(self, d: List[List[int]]) -> int:
        m=len(d)
        n=len(d[0])
        dp=[[0]*n for _ in range(m)]
        #base case 
        if d[m-1][n-1]<0:
            dp[m-1][n-1]=abs(d[m-1][n-1])+1
        else:
            dp[m-1][n-1]=1
        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                if i==m-1 and j==n-1:
                    continue
                if i==m-1:
                    if d[i][j]>0:
                        if d[i][j]>=dp[i][j+1]:
                            dp[i][j]=1
                        else:
                            dp[i][j]=dp[i][j+1]-d[i][j]
                    else:
                        dp[i][j]= abs(d[i][j])+dp[i][j+1]
                if j==n-1:
                    if d[i][j]>0:
                        if d[i][j]>=dp[i+1][j]:
                            dp[i][j]=1
                        else:
                            dp[i][j]=dp[i+1][j]-d[i][j]
                    else:
                        dp[i][j]= abs(d[i][j])+dp[i+1][j]
                elif i<m-1 and j<n-1:
                    mini=min(dp[i][j+1],dp[i+1][j])
                    if d[i][j]<=0:
                        dp[i][j]=abs(d[i][j])+mini
                    else:
                        if d[i][j]<mini:
                            dp[i][j]=mini-d[i][j]
                        else:
                            dp[i][j]=1 
        return dp[0][0]

        
        

File: 2024-4-17/dungeon game.js
/**
 * @param {number[][]} dungeon
 * @return {number}
 */
var calculateMinimumHP = function(dungeon) {
    const n = dungeon.length;
    const m = dungeon[0].length;

    const dp = new Array(n).fill(0).map(() => new Array(m).fill(0));

    return helper(dungeon, n, m, 0, 0, dp);
};

var helper = function(dungeon, n, m, row, col, dp) {
    // base condition 1
    if (row === n - 1 && col === m - 1) {
        return Math.max(1, 1 - dungeon[row][col]);
    }

    // base condition 2
    if (row >= n || col >= m) {
        return Infinity;
    }

    if (dp[row][col] !== 0) {
        return dp[row][col];
    }

    const right = helper(dungeon, n, m, row, col + 1, dp);
    const down = helper(dungeon, n, m, row + 1, col, dp);

    // Calculate the minimum health needed at the current position
    dp[row][col] = Math.max(1, Math.min(right, down) - dungeon[row][col]);

    return dp[row][col];
};